PATHFIND/GetVehicleNodeProperties.md

```diff
@@ -11,7 +11,22 @@ BOOL GET_VEHICLE_NODE_PROPERTIES(float x, float y, float z, int* density, int* f
 ```
 Gets the density and flags of the closest node to the specified position.  
 Density is a value between 0 and 15, indicating how busy the road is.  
-Flags is a bit field.  
+
+Flags is a bit field:
+public enum PathNodeFlags
+{
+    Offroad = 0x1,
+    WanderTarget = 0x2,
+    NoBigVehicles = 0x4,
+    Disabled = 0x8,
+    Tunnel = 0x10,
```

Comment:
Same thing as above, this shouldn't be in a code block

Reply:
> I don't have much of `CPathFind` documented, so these are my _rough_ notes on this batch of natives. Unsure how valuable they will be to continue your research:
> 
> #### 0x132F52BBA570FE92:
> `p3` looks like a node radius and is used to supplement `n`, likely leading to your "A bit broken as it returns a random edge after `n`" comment. For example, and using your naming convention, here is an outline of one of the distance checks:
> 
> ```c
> if ((nodeForwardLanes + nodeBackwardLanes < n) || distance(node, linkedNode) <= p3)
>     closestDistance = rootManhattanDistance
> else {
>     closestDistance = nodeManhattanDistance // (dy + dx) + (dz * 3.0)
>     ...
> }
> ```
> 
> #### 0x80CA6A8B6C094CC4:
> `nodeType` looks like a bitfield/flag not an explicit type.
> 
> #### 0x0568566ACBB5DEDC:
> Some additional things to check:
> 
> ```
> 0x10  -> Also used in interiors; 'Tunnels' may be a specific instance of this flag.
> 0x20  -> Maybe a combination of "Block If No Lanes" and "Narrowroad"? Grep script usage: iVar4 & 32 == 0 && iVar4 & 4 == 0?
> 0x100 -> Stop/TrafficLightStop
> 0x200 -> Intersection/Crosswalk
> 0x400 -> Water
> ```
> 
> If you are using Codewalker, here are the `YndNode` mappings (I think):
> 
> ```cs
> 0x1 = ((RawData.Flags0 & 8) != 0)
> 0x2 = ((RawData.Flags0 & 0x10) != 0)
> 0x4 = ((RawData.Flags0 & 0x20) != 0)
> 0x8 = ((RawData.Flags2 & 0x80) != 0)
> 0x10 = ((RawData.Flags3 & 1) != 0)
> 0x20 = ((RawData.Flags4 & 0x70) != 0)
> 0x40 = ((RawData.Flags2 & 0x40) != 0 && (RawData.Flags2 & 0x20) == 0)
> 0x80 = ((RawData.LinkCountFlags & 0x1) != 0)
> 0x100 = ((RawData.Flags1 & 0xF8) == 0x78)
> 0x200 = ((RawData.Flags1 & 0xF8) == 0x80)
> 0x400 = ((RawData.Flags2 & 0x20) != 0)
> ```

Thanks for replying. I've actually done a fair bit of extensive research on them that was never merged (need to touch base with dexy to figure out where it is). It was all figured out by cross referencing the incomplete `paths.xml` so the flags names should be accurate. `Junction` is also `Flags2 & 4`. I've yet to push that up.

https://github.com/dexyfex/CodeWalker/pull/70/files